### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by SqREL**

### CODE REVIEW

File: app/controllers/api/admin_controller.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
+class Api::AdminController < Api::ApplicationController
+
+  def initialize
+    @max_rent_time = 15
+    @rental_cost = 1
+    @compensation_cost = 0.5
```
Comment (SqREL) [inline app/controllers/api/admin_controller.rb]: I don't understand why this values are instance vars, not constants. By fact - they ARE constants, but somewhy are initialised in constructor every time controller's instance created.

File: app/controllers/api/admin_controller.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  end
+
+  # GET /api/admin/stock
+  def stock
+    rented_cards = Card.where('rent_time > ?', Time.now - @max_rent_time.minutes).count
+    lost_cards = Card.where('rent_time <= ?', Time.now - @max_rent_time.minutes).count
+    state = session[:state_cards].present? ? session[:state_cards] : 53
```
Comment (SqREL) [inline app/controllers/api/admin_controller.rb]: 1. how sessions will work here, if it's an api? So client should use extra headers to see correct result?
2. So if someone, from other session will "get a card" - he/she will see a new card deck (53 cards)?

File: app/controllers/api/admin_controller.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  def initialize
+    @max_rent_time = 15
+    @rental_cost = 1
+    @compensation_cost = 0.5
+  end
+
+  # GET /api/admin/stock
+  def stock
+    rented_cards = Card.where('rent_time > ?', Time.now - @max_rent_time.minutes).count
+    lost_cards = Card.where('rent_time <= ?', Time.now - @max_rent_time.minutes).count
```
Comment (SqREL) [inline app/controllers/api/admin_controller.rb]: so here you have a deal with only cards count. But in test assignment it says, that cards are not "noname", they have their values: 

> Permitted suit values are ["spade", "heart", "club", "diamond"] Permitted card values are "2" to "13" and "a". (A is for Ace) The exception here is the joker, which has a null suit and a card value of "joker".

File: app/controllers/api/admin_controller.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    render json: {:available_cards => available_cards, :rented_cards => rented_cards, :lost_cards => lost_cards}
+  end
+
+  # GET /api/admin/restock
+  def restock
+    cards = Card.where('rent_time < ?', Time.now - @max_rent_time.minutes)
+
+    compensation_amount = (cards.count * @compensation_cost)
+    session[:balance] ||= 5
+    session[:balance] -= compensation_amount
```
Comment (SqREL) [inline app/controllers/api/admin_controller.rb]: Why sessions?

File: app/controllers/api/admin_controller.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    lost_cards = Card.where('rent_time <= ?', Time.now - @max_rent_time.minutes).count
+    state = session[:state_cards].present? ? session[:state_cards] : 53
+    available_cards = state - rented_cards - lost_cards
+
+    render json: {:available_cards => available_cards, :rented_cards => rented_cards, :lost_cards => lost_cards}
+  end
+
+  # GET /api/admin/restock
+  def restock
+    cards = Card.where('rent_time < ?', Time.now - @max_rent_time.minutes)
```
Comment (SqREL) [inline app/controllers/api/admin_controller.rb]: i'm trying to understand what exactly `cards` do you mean here. It looks like they are `lost_cards` from the line 12, but i can't get why you use `<` here, but `<=` there. Anyway the naming is controversial.

File: app/controllers/api/card_controller.rb [hunk 6]
```diff
@@ -0,0 +1,63 @@
+class Api::CardController < Api::ApplicationController
+
+  protect_from_forgery with: :null_session, if: Proc.new { |c| c.request.format == 'application/json' }
+
+  def initialize
```
Comment (juanpedrojose) [inline app/controllers/api/card_controller.rb]: Same here.
You are monkeypatching your controller's initializer and even more - without calling `super`.
Such solution is not required by this assignment and besides is not a rails-way.

File: app/controllers/api/admin_controller.rb [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+      render json: {'message' => 'The cards have been restoked', :lost_compensation => compensation_amount}
+    else
+      render json: {'message' => 'Somethin went wrong'}
+    end
+  end
+
+  # GET /api/admin/finances
+  def finances
+    pending_rent = 0.0
+    rented_cards = Card.where('rent_time > ?', Time.now - @max_rent_time.minutes)
```
Comment (SqREL) [inline app/controllers/api/admin_controller.rb]: it's the same line as on line 11

File: config/database.yml [hunk 8]
```diff
@@ -1,25 +1,25 @@
-# SQLite version 3.x
-#   gem install sqlite3
-#
-#   Ensure the SQLite 3 gem is defined in your Gemfile
-#   gem 'sqlite3'
-#
 default: &default
-  adapter: sqlite3
+  adapter: mysql2
```
Comment (juanpedrojose) [inline config/database.yml]: You have switched from sqlite to mysql.
Such solution will require:
- an installed mysql server;
- a configured user and database;

what are the benefits of such solution over sqlite?

File: log/development.log [hunk 9]
```diff
@@ -0,0 +1,795 @@
+  [1m[36mActiveRecord::SchemaMigration Load (0.2ms)[0m  [1mSELECT `schema_migrations`.* FROM `schema_migrations`[0m
```
Comment (juanpedrojose) [inline log/development.log]: Please do not include log files in your commits.
This basically just spam.

File: log/test.log [hunk 10]
```diff
@@ -0,0 +1,991 @@
+  [1m[36m (83.2ms)[0m  [1mDROP DATABASE IF EXISTS `denmark_test`[0m
```
Comment (juanpedrojose) [inline log/test.log]: And again, no log files.

File: app/controllers/api/admin_controller.rb [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+      card.in_use = nil
+      card.is_lost = nil
+      card.rent_time = nil
+      card.save
+    end
+
+    unless session[:balance] < 0
+      render json: {'message' => 'The cards have been restoked', :lost_compensation => compensation_amount}
+    else
+      render json: {'message' => 'Somethin went wrong'}
```
Comment (SqREL) [inline app/controllers/api/admin_controller.rb]: but it will return `200 OK`. It doens't meet requirements

File: spec/factories/cards.rb [hunk 12]
```diff
@@ -0,0 +1,9 @@
+FactoryGirl.define do
+
+  factory :card do
```
Comment (juanpedrojose) [inline spec/factories/cards.rb]: This factory looks peculiar.

File: app/controllers/api/admin_controller.rb (hunk 3 above)
Comment (SqREL) [inline app/controllers/api/admin_controller.rb]: also there's another reason why it won't meet requirements:

> When the app is launched, it launches with a standard deck of cards, which is 52 cards + 1 joker in the inventory.

but instead previous state will be saved in this solution

File: app/controllers/api/api_application_controller.rb [hunk 13]
```diff
@@ -0,0 +1,5 @@
+class Api::ApplicationController < ActionController::Base
+  # Prevent CSRF attacks by raising an exception.
+  # For APIs, you may want to use :null_session instead.
+  protect_from_forgery with: :exception
```
Comment (SqREL) [inline app/controllers/api/api_application_controller.rb]: why you need protection from csrf in api?

File: app/controllers/api/card_controller.rb [hunk 14]
```diff
@@ -0,0 +1,63 @@
+class Api::CardController < Api::ApplicationController
+
+  protect_from_forgery with: :null_session, if: Proc.new { |c| c.request.format == 'application/json' }
```
Comment (SqREL) [inline app/controllers/api/card_controller.rb]: well no.
1. why null session?
2. why session in api? (again)

> We want to cash in on the new digital content market, and therefore we are launching a new app where you can rent virtual gaming cards using a JSON HTTP API

so ideally it should return 406 or something

File: app/controllers/api/card_controller.rb [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
+  def initialize
+    @max_rent_time = 15
+    @rental_cost = 1
+  end
+
+  # GET /api/launch
+  def launch
```
Comment (SqREL) [inline app/controllers/api/card_controller.rb]: no such endpoint in requirements

i cannot understand at all how you can launch an application by calling endpoint of the same application (that hadn't been even lauched)

File: app/controllers/api/card_controller.rb [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+    response = {
+      :message => 'Application has launched',
+      :initial_balance => session[:balance],
+      :state_cards => session[:state_cards]
+    }
+
+    render json: response
+  end
+
+  # GET /api/cards
```
Comment (SqREL) [inline app/controllers/api/card_controller.rb]: i thought it's a typo or something, but it turns out, that it is true.
Why do you call GET on /api/cards and call `create` method?

File: app/controllers/api/card_controller.rb [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+  def index
+    render json: Card.all
+  end
+
+  # GET /api/card
+  def create
```
Comment (SqREL) [inline app/controllers/api/card_controller.rb]: i thought it's a typo or something, but it turns out, that it is true.
Why do you call GET on /api/cards and call create method?

File: app/controllers/api/card_controller.rb [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+    card = Card.where(in_use: nil).order("RAND()").last
```
Comment (SqREL) [inline app/controllers/api/card_controller.rb]: sarch & update should be done in a transaction in order of data consistency

File: app/controllers/api/card_controller.rb [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+      end
+    else
+      render json: {'error' => 'Cards not found'}, status: 404
+    end
+  end
+
+  # PUT /api/card
+  def update
+    user_id = 1 # only for testing purposes, in a real projects user id must be taken from the auth session
+    user = Client.find(user_id)
```
Comment (SqREL) [inline app/controllers/api/card_controller.rb]: but you have asked about users, and i told you, that there's no users at all.
I don't understand why there should be users at all

File: app/controllers/api/card_controller.rb [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+    card = Card.where(in_use: true).find_by(id: params[:id])
+    if user.present? && card.present?
+      rent_time = ((Time.now - card.rent_time)/60).ceil
+      unless (rent_time >= @max_rent_time.minutes)
+        card.update_attributes(:in_use => nil, :rent_time => nil)
+        amount = (rent_time * @rental_cost).round(2)
+        Transaction.create(:user_id => user.id, :amount => amount, :created_at => Time.now)
```
Comment (SqREL) [inline app/controllers/api/card_controller.rb]: should be wrapped into transactions in order of data consistency

File: app/helpers/api_helper.rb [hunk 21]
```diff
@@ -0,0 +1,2 @@
+module ApiHelper
```
Comment (SqREL) [inline app/helpers/api_helper.rb]: useless

File: app/models/card.rb [hunk 22]
```diff
@@ -0,0 +1,2 @@
+class Card < ActiveRecord::Base
+end
```
Comment (SqREL) [inline app/models/card.rb]: some logic could be here, instead of overloading controllers

File: app/controllers/api/card_controller.rb [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+  end
+
+  # GET /api/card
+  def create
+    card = Card.where(in_use: nil).order("RAND()").last
+    if card.present?
+      if card.update_attributes(in_use: true, rent_time: Time.now)
+        render json: card
+      else
+        render json: {'error' => 'Cards can not be updated'}, status: 404
```
Comment (SqREL) [inline app/controllers/api/card_controller.rb]: 404 means `not found`. But the card is found.

**Pull request closed without merging**